File: src/lib/manifest/manifestService.ts
import fs from "node:fs/promises";
import path from "node:path";

import { getBungieApiKey } from "../config/env";
import {
  currentVersionFilePath,
  entityCacheMetaPath,
  rawTablePath,
} from "./cachePaths";
import {
  RAW_TABLES,
  type ManifestService,
  type ManifestStatus,
  type RawTable,
  type RawTableName,
} from "./types/services";
import type { EntityCacheMeta } from "./types/stores";

const MANIFEST_URL = "https://www.bungie.net/Platform/Destiny2/Manifest/";
const BUNGIE_BASE_URL = "https://www.bungie.net";
const MISSING_API_KEY_MESSAGE =
  "BUNGIE_API_KEY environment variable is required to download manifest data";

interface BungieManifestServiceOptions {
  apiKey: string | null;
  fetchFn?: typeof fetch;
}

interface ManifestMetadata {
  version: string;
  tablePaths: Record<string, string>;
}

interface CurrentVersionFile {
  version: string;
}

function isNodeError(error: unknown): error is NodeJS.ErrnoException {
  return error instanceof Error && "code" in error;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isRawTable(value: unknown): value is RawTable {
  return isRecord(value);
}

function isCurrentVersionFile(value: unknown): value is CurrentVersionFile {
  return isRecord(value) && typeof value.version === "string";
}

function isEntityCacheMeta(value: unknown): value is EntityCacheMeta {
  if (!isRecord(value)) return false;
  if (typeof value.manifestVersion !== "string") return false;
  if (typeof value.builtAt !== "string") return false;
  return isRecord(value.counts);
}

function parseManifestResponse(json: unknown): ManifestMetadata {
  if (!isRecord(json) || !isRecord(json.Response)) {
    throw new Error("Unexpected Bungie manifest response shape");
  }

  const response = json.Response;
  if (typeof response.version !== "string") {
    throw new Error("Manifest response missing version");
  }

  const pathsRoot = response.jsonWorldComponentContentPaths;
  if (!isRecord(pathsRoot) || !isRecord(pathsRoot.en)) {
    throw new Error("Manifest response missing English table paths");
  }

  const tablePaths: Record<string, string> = {};
  for (const [tableName, tablePath] of Object.entries(pathsRoot.en)) {
    if (typeof tablePath === "string") {
      tablePaths[tableName] = tablePath;
    }
  }

  return { version: response.version, tablePaths };
}

async function readJsonFile(filePath: string): Promise<unknown | null> {
  try {
    const content = await fs.readFile(filePath, "utf-8");
    return JSON.parse(content) as unknown;
  } catch (error) {
    if (isNodeError(error) && error.code === "ENOENT") {
      return null;
    }
    return null;
  }
}

async function writeJsonFile(filePath: string, data: unknown): Promise<void> {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, JSON.stringify(data), "utf-8");
}

async function fileExists(filePath: string): Promise<boolean> {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function readCurrentVersion(): Promise<string | null> {
  const json = await readJsonFile(currentVersionFilePath());
  if (!isCurrentVersionFile(json)) {
    return null;
  }
  return json.version;
}

async function readEntityCacheMeta(
  version: string,
): Promise<EntityCacheMeta | null> {
  const json = await readJsonFile(entityCacheMetaPath(version));
  if (!isEntityCacheMeta(json)) {
    return null;
  }
  return json;
}

function computeIsStale(
  cachedVersion: string | null,
  remoteVersion: string | null,
): boolean {
  if (cachedVersion === null) {
    return true;
  }
  if (remoteVersion === null) {
    return false;
  }
  return cachedVersion !== remoteVersion;
}

export class BungieManifestService implements ManifestService {
  private readonly apiKey: string | null;
  private readonly fetchFn: typeof fetch;

  constructor(options: BungieManifestServiceOptions) {
    this.apiKey = options.apiKey;
    this.fetchFn = options.fetchFn ?? fetch;
  }

  async getStatus(): Promise<ManifestStatus> {
    const cachedVersion = await readCurrentVersion();
    const remoteVersion = await this.fetchRemoteVersion();
    const entityCache =
      cachedVersion === null
        ? null
        : await readEntityCacheMeta(cachedVersion);

    return {
      cachedVersion,
      remoteVersion,
      isStale: computeIsStale(cachedVersion, remoteVersion),
      entityCache,
    };
  }

  async ensureCurrent(): Promise<string> {
    if (this.apiKey === null) {
      throw new Error(MISSING_API_KEY_MESSAGE);
    }

    const metadata = await this.fetchManifestMetadata(this.apiKey);
    await this.downloadMissingTables(metadata);
    await writeJsonFile(currentVersionFilePath(), {
      version: metadata.version,
    });
    return metadata.version;
  }

  async loadRawTable(
    version: string,
    table: RawTableName,
  ): Promise<RawTable> {
    const filePath = rawTablePath(version, table);

    let content: string;
    try {
      content = await fs.readFile(filePath, "utf-8");
    } catch (error) {
      if (isNodeError(error) && error.code === "ENOENT") {
        throw new Error(
          `Raw table "${table}" for version "${version}" is not on disk. Call ensureCurrent() to download manifest tables.`,
        );
      }
      throw error;
    }

    const parsed: unknown = JSON.parse(content);
    if (!isRawTable(parsed)) {
      throw new Error(
        `Raw table "${table}" for version "${version}" is not a valid JSON object.`,
      );
    }
    return parsed;
  }

  private async fetchRemoteVersion(): Promise<string | null> {
    if (this.apiKey === null) {
      return null;
    }

    try {
      const metadata = await this.fetchManifestMetadata(this.apiKey);
      return metadata.version;
    } catch {
      return null;
    }
  }

  private async fetchManifestMetadata(apiKey: string): Promise<ManifestMetadata> {
    const response = await this.fetchFn(MANIFEST_URL, {
      headers: { "X-API-Key": apiKey },
    });

    if (!response.ok) {
      throw new Error(`Manifest version check failed with status ${response.status}`);
    }

    const json: unknown = await response.json();
    return parseManifestResponse(json);
  }

  private async downloadMissingTables(metadata: ManifestMetadata): Promise<void> {
    const apiKey = this.apiKey;
    if (apiKey === null) {
      throw new Error(MISSING_API_KEY_MESSAGE);
    }

    for (const table of RAW_TABLES) {
      const destination = rawTablePath(metadata.version, table);
      if (await fileExists(destination)) {
        continue;
      }

      await this.downloadTable(apiKey, metadata, table, destination);
    }
  }

  private async downloadTable(
    apiKey: string,
    metadata: ManifestMetadata,
    table: RawTableName,
    destination: string,
  ): Promise<void> {
    const relativePath = metadata.tablePaths[table];
    if (!relativePath) {
      throw new Error(`Manifest response missing download path for ${table}`);
    }

    const url = `${BUNGIE_BASE_URL}${relativePath}`;
    const response = await this.fetchFn(url, {
      headers: { "X-API-Key": apiKey },
    });

    if (!response.ok) {
      throw new Error(
        `Failed to download ${table}: HTTP ${response.status}`,
      );
    }

    const json: unknown = await response.json();
    if (!isRawTable(json)) {
      throw new Error(`Downloaded ${table} is not a valid JSON object`);
    }

    await writeJsonFile(destination, json);
  }
}

export function createManifestService(): ManifestService {
  return new BungieManifestService({ apiKey: getBungieApiKey() });
}

File: flutter/packages/manifest/lib/src/manifest_service.dart
import 'dart:convert';

import 'package:destiny2_storage/destiny2_storage.dart';

import 'http_client.dart';
import 'io/text_file.dart' as text_file;
import 'types/services.dart';
import 'types/stores.dart';

const kManifestUrl = 'https://www.bungie.net/Platform/Destiny2/Manifest/';
const kBungieBaseUrl = 'https://www.bungie.net';

const kMissingApiKeyMessage =
    'BUNGIE_API_KEY is required to download manifest data';

/// Stage 1: versioned download of raw manifest tables under [StorageRoot].
///
/// Port of product `BungieManifestService` for Windows / pure Dart hosts
/// (DART-018). Paths use [StorageRoot], not repo `.cache`.
/// File IO is conditional — web uses prebuilt entity bundles (DART-044).
class BungieManifestService {
  BungieManifestService({
    required this.storageRoot,
    required this.apiKey,
    ManifestHttpGet? httpGet,
    this.cacheRawTables = true,
    List<String>? tablesToDownload,
  })  : httpGet = httpGet ?? createUtf8ManifestHttpGet(),
        tablesToDownload = tablesToDownload ?? downloadRawTables;

  final StorageRoot storageRoot;

  /// Public Bungie API key only — never a client secret.
  final String? apiKey;

  final ManifestHttpGet httpGet;

  /// In-process memo of parsed raw tables.
  final bool cacheRawTables;

  /// Tables downloaded by [ensureCurrent] (default: product [downloadRawTables]).
  final List<String> tablesToDownload;

  final Map<String, Future<RawTable>> _rawTableCache = {};

  /// Settings-friendly status: cached / remote / stale / entity meta.
  Future<ManifestStatus> getStatus() async {
    final cachedVersion = await readCurrentVersion();
    final remoteVersion = await fetchRemoteVersion();
    final entityCache = cachedVersion == null
        ? null
        : await readEntityCacheMeta(cachedVersion);

    return ManifestStatus(
      cachedVersion: cachedVersion,
      remoteVersion: remoteVersion,
      isStale: computeIsStale(cachedVersion, remoteVersion),
      entityCache: entityCache,
    );
  }

  /// Downloads tables for the latest remote version.
  Future<String> ensureCurrent({bool forceFullDownload = false}) async {
    final key = apiKey;
    if (key == null || key.isEmpty) {
      throw ManifestServiceException(kMissingApiKeyMessage);
    }

    final metadata = await fetchManifestMetadata(key);
    await downloadTables(
      metadata,
      forceFullDownload: forceFullDownload,
    );
    await writeCurrentVersion(metadata.version);
    _rawTableCache.clear();
    return metadata.version;
  }

  /// Loads one raw table for a downloaded [version] from disk.
  Future<RawTable> loadRawTable(String version, String table) async {
    final cacheKey = '$version::$table';
    if (cacheRawTables) {
      final hit = _rawTableCache[cacheKey];
      if (hit != null) return hit;
    }

    final loadPromise = _readAndParseRawTable(version, table);
    if (cacheRawTables) {
      _rawTableCache[cacheKey] = loadPromise;
      loadPromise.then<void>(
        (_) {},
        onError: (Object _, StackTrace __) {
          if (identical(_rawTableCache[cacheKey], loadPromise)) {
            _rawTableCache.remove(cacheKey);
          }
        },
      );
    }
    return loadPromise;
  }

  Future<RawTable> _readAndParseRawTable(String version, String table) async {
    final filePath = storageRoot.rawTablePath(version, table);
    final content = await text_file.readTextFile(filePath);
    if (content == null) {
      throw ManifestServiceException(
        'Raw table "$table" for version "$version" is not on disk. '
        'Call ensureCurrent() to download manifest tables.',
      );
    }
    final parsed = jsonDecode(content);
    if (parsed is! Map) {
      throw ManifestServiceException(
        'Raw table "$table" for version "$version" is not a valid JSON object.',
      );
    }
    // Prefer the decoded map as-is — Map.from on DestinyInventoryItemDefinition
    // (~190MB JSON) roughly doubles peak memory and can OOM Windows hosts.
    if (parsed is Map<String, dynamic>) return parsed;
    return Map<String, dynamic>.from(parsed);
  }

  Future<String?> readCurrentVersion() async {
    final text = await text_file.readTextFile(storageRoot.currentVersionPath);
    if (text == null) return null;
    try {
      final json = jsonDecode(text);
      if (json is Map && json['version'] is String) {
        return json['version'] as String;
      }
    } catch (_) {
      return null;
    }
    return null;
  }

  Future<void> writeCurrentVersion(String version) async {
    await text_file.writeTextFile(
      storageRoot.currentVersionPath,
      jsonEncode({'version': version}),
    );
  }

  Future<EntityCacheMeta?> readEntityCacheMeta(String version) async {
    final text =
        await text_file.readTextFile(storageRoot.entityCacheMetaPath(version));
    if (text == null) return null;
    try {
      return EntityCacheMeta.fromJson(
        Map<String, dynamic>.from(jsonDecode(text) as Map),
      );
    } catch (_) {
      return null;
    }
  }

  Future<String?> fetchRemoteVersion() async {
    final key = apiKey;
    if (key == null || key.isEmpty) return null;
    try {
      final metadata = await fetchManifestMetadata(key);
      return metadata.version;
    } catch (_) {
      return null;
    }
  }

  Future<ManifestMetadata> fetchManifestMetadata(String apiKey) async {
    final response = await httpGet(
      Uri.parse(kManifestUrl),
      headers: {'X-API-Key': apiKey},
    );
    if (!response.ok) {
      throw ManifestServiceException(
        'Manifest version check failed with status ${response.statusCode}',
      );
    }
    return parseManifestResponse(jsonDecode(response.body));
  }

  Future<void> downloadTables(
    ManifestMetadata metadata, {
    required bool forceFullDownload,
  }) async {
    final key = apiKey;
    if (key == null || key.isEmpty) {
      throw ManifestServiceException(kMissingApiKeyMessage);
    }

    for (final table in tablesToDownload) {
      final destination = storageRoot.rawTablePath(metadata.version, table);
      if (!forceFullDownload && await text_file.textFileExists(destination)) {
        continue;
      }
      await downloadTable(key, metadata, table, destination);
    }
  }

  Future<void> downloadTable(
    String apiKey,
    ManifestMetadata metadata,
    String table,
    String destination,
  ) async {
    final relativePath = metadata.tablePaths[table];
    if (relativePath == null || relativePath.isEmpty) {
      throw ManifestServiceException(
        'Manifest response missing download path for $table',
      );
    }

    final url = Uri.parse('$kBungieBaseUrl$relativePath');
    final response = await httpGet(
      url,
      headers: {'X-API-Key': apiKey},
    );
    if (!response.ok) {
      throw ManifestServiceException(
        'Failed to download $table: HTTP ${response.statusCode}',
      );
    }

    final parsed = jsonDecode(response.body);
    if (parsed is! Map) {
      throw ManifestServiceException(
        'Downloaded $table is not a valid JSON object',
      );
    }

    await text_file.writeTextFile(destination, jsonEncode(parsed));
  }
}

/// Parse Bungie Platform Destiny2/Manifest JSON body.
ManifestMetadata parseManifestResponse(Object? json) {
  if (json is! Map) {
    throw ManifestServiceException('Unexpected Bungie manifest response shape');
  }
  final response = json['Response'];
  if (response is! Map) {
    throw ManifestServiceException('Unexpected Bungie manifest response shape');
  }
  final version = response['version'];
  if (version is! String) {
    throw ManifestServiceException('Manifest response missing version');
  }

  final pathsRoot = response['jsonWorldComponentContentPaths'];
  if (pathsRoot is! Map) {
    throw ManifestServiceException(
      'Manifest response missing English table paths',
    );
  }
  final en = pathsRoot['en'];
  if (en is! Map) {
    throw ManifestServiceException(
      'Manifest response missing English table paths',
    );
  }

  final tablePaths = <String, String>{};
  for (final entry in en.entries) {
    if (entry.value is String) {
      tablePaths[entry.key.toString()] = entry.value as String;
    }
  }

  return ManifestMetadata(version: version, tablePaths: tablePaths);
}
